### ISSUE TITLE
Can't reproduce on MacOSX

### ISSUE THREAD

**Issue by gnzlbg** [glandium/global_allocator_linkage-rs#1]
On MacOSX, running `make` I do get the expected linker error about duplicate symbols, which avoids the undefined behavior you are seeing: 

```
make
clang -o foobar.o -c foobar.c -O2
cd foo && cargo build --release
    Updating crates.io index
   Compiling libc v0.2.60
   Compiling foo v0.1.0 (/Users/gnzlbg/projects/sideprojects/global_allocator_linkage-rs/foo)
    Finished release [optimized] target(s) in 10.12s
cd bar && cargo build --release
    Updating crates.io index
   Compiling libc v0.2.60
   Compiling bar v0.1.0 (/Users/gnzlbg/projects/sideprojects/global_allocator_linkage-rs/bar)
    Finished release [optimized] target(s) in 2.90s
clang -o foobar foobar.o foo/target/release/libfoo.a bar/target/release/libbar.a -lpthread -ldl
duplicate symbol ___rg_alloc_zeroed in:
    bar/target/release/libbar.a(bar-53f9af0b5a20a35b.bar.4ouxddql-cgu.1.rcgu.o)
    foo/target/release/libfoo.a(foo-a766069b19f4385e.foo.eag3z2tp-cgu.1.rcgu.o)
duplicate symbol ___rg_realloc in:
    bar/target/release/libbar.a(bar-53f9af0b5a20a35b.bar.4ouxddql-cgu.1.rcgu.o)
    foo/target/release/libfoo.a(foo-a766069b19f4385e.foo.eag3z2tp-cgu.1.rcgu.o)
duplicate symbol ___rg_dealloc in:
    bar/target/release/libbar.a(bar-53f9af0b5a20a35b.bar.4ouxddql-cgu.1.rcgu.o)
    foo/target/release/libfoo.a(foo-a766069b19f4385e.foo.eag3z2tp-cgu.1.rcgu.o)
duplicate symbol ___rg_alloc in:
    bar/target/release/libbar.a(bar-53f9af0b5a20a35b.bar.4ouxddql-cgu.1.rcgu.o)
    foo/target/release/libfoo.a(foo-a766069b19f4385e.foo.eag3z2tp-cgu.1.rcgu.o)
ld: 4 duplicate symbols for architecture x86_64
clang-8: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [foobar] Error 1
```

For a bug report to rust-lang/rust, please include which target, C compilers, linker, etc. you are using.

**Comment by glandium** [comment]
I don't think it's fixable, short of killing reproducibility. It's due to the way ld works in ELF world.

**Comment by glandium** [comment]
(although I must say, I was expecting it to fail on mac too... maybe it'd fail similarly on windows)

**Comment by gnzlbg** [comment]
> It's due to the way ld works in ELF world.

I don't know any details about linking, but in C you have `inline` and non-`inline` functions, and the main difference is that the symbols of `inline` functions are merged, while the symbols of non-`inline` functions cannot. It would surprise me if `ld` on Linux wouldn't support that, and IIRC I often saw errors due to duplicate symbols when using C++ on Linux (e.g. because I forgot to make a function in a header file `inline`).

We do hit this on Linux, e.g., for the musl targets, where libstd links musl statically, and if, e.g., libc from crates.io also does that, then musl is linked statically twice, and the binary fails to link with ld due to duplicated symbols.

So I think there must be a way to make ld fail for this case on Linux as well.

**Comment by glandium** [comment]
Even without global_allocator, if you do the same with two crates that just have a function that `println!("foo")` each, both static libraries will contain a copy of libstd, and everything will compile and link fine.

**Comment by gnzlbg** [comment]
> Even without global_allocator, if you do the same with two crates that just have a function that println!("foo") each, both static libraries will contain a copy of libstd, and everything will compile and link fine.

Does this work for `x86_64-unknown-linux-musl` ?

**Comment by glandium** [comment]
I don't have a x86_64-unknown-linux-musl toolchain.

**Comment by gnzlbg** [comment]
IIRC, it doesn't work for that toolchain, because the C library is statically linked.

It works for `x86_64-unknown-linux-gnu` because glibc is dynamically linked by default.

I just think that when we generate the linker symbols for the allocation shims that are used to call the `#[global_allocator]` functions, we need to tell the linker that duplicates of these symbols cannot be merged.

**Comment by glandium** [comment]
FWIW, the test case works as is for me on mac.

**Comment by gnzlbg** [comment]
By "works", do you mean that it links successfully and has undefined behavior without linker errors, or do you get linker errors ?

**Comment by glandium** [comment]
I mean it has the same behavior as on Linux on my mac.

**Comment by glandium** [comment]
Do you have RUSTFLAGS set to enable LTO?

**Comment by gnzlbg** [comment]
> I mean it has the same behavior as on Linux on my mac.

I see, so you can't reproduce this issue there. Weird.

> Do you have RUSTFLAGS set to enable LTO?

Nope, I don't have anything. I'm using:

```
rustc 1.38.0-nightly (8a58268b5 2019-07-31)
```

and I can reproduce this with both Apple Clang:

```
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

and clang 8:

```
clang version 8.0.0 (git@github.com:llvm-mirror/clang.git 1bc73590ad1335313e8f262393547b8af67c9167) (git@github.com:llvm-mirror/llvm.git 24b52d3cc6d314ff13a05b1112871f22fa063ab7)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Users/gnzlbg/pool/toolchain/llvm/bin
```

I'm using the default `ld` in both cases:

```
@(#)PROGRAM:ld  PROJECT:ld64-450.3
BUILD 18:45:16 Apr  4 2019
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 10.0.1, (clang-1001.0.46.4) (static support for 22, runtime is 22)
TAPI support using: Apple TAPI version 10.0.1 (tapi-1001.0.4.1)
```

**Comment by glandium** [comment]
Interesting, I can reproduce with rustc nightly. But not on Linux.

**Comment by gnzlbg** [comment]
I think this is worth opening an issue in rust-lang/rust. I think this should always produce a linker error.

**Comment by glandium** [comment]
Do you think https://github.com/glandium/global_allocator_linkage-rs/issues/1#issuecomment-519433195 should too? Because I sure don't think it should.

**Comment by gnzlbg** [comment]
No that comment isn't really related. You tested that with `x86_64-unknown-linux-gnu`, and for that target, that's expected and guaranteed to work (unless one messes with `crt-static`), since the symbols are dynamically linked, so you can link them as many times as you want.

**Comment by glandium** [comment]
I'm more talking about the "conflicting" symbols for libstd.

**Comment by gnzlbg** [comment]
> I'm more talking about the "conflicting" symbols for libstd.

Ah, that problem isn't new. There are a couple of issues open for them for the targets in which the C library is statically linked (e.g. the musl ones).

**Comment by glandium** [comment]
I mean, on Linux (glibc) and mac, two rust staticlib's libstd don't conflict. Would you rather they do? (I wouldn't)

**Comment by gnzlbg** [comment]
> I mean, on Linux (glibc) and mac, two rust staticlib's libstd don't conflict. Would you rather they do? 

No, that's the intended behavior. They should only conflict if you put a `#[global_allocator]` (or something similar to that) on each, because then we don't which one to use.

**Comment by glandium** [comment]
What about if one has a #[global_allocator] and the other not? (although I guess in that case it means an implicit one?)

**Comment by glandium** [comment]
What if they have one each, but from the same crate, built the same way?

**Comment by gnzlbg** [comment]
I'm not sure that you have understood what `#[global_allocator]` does. It defines a global singleton. If there is more than one of those anywhere in a binary, the behavior is undefined. 

That is, anything can happen, including that one of them is picked at random, that your program fails to link, to compile, that cargo errors, or that running the binary eats your laundry. 

> What about if one has a #[global_allocator] and the other not?

If the final binary only contains at most one `#[global_allocator]` this is fine.

>  (although I guess in that case it means an implicit one?)

No, if there is no `#[global_allocator]` then there is no `#[global_allocator]` - that's all there is to it.

> What if they have one each, but from the same crate, built the same way?

It doesn't matter that two `#[global_allocator]`s are the same. If there are more than one, the behavior is undefined. 

---

That is, if you put a `#[global_allocator]` into a library, then that library cannot be used with any other code that does the same without invoking undefined behavior. If you are able to do this with safe Rust code, then this might be a soundness bug.

The only sane way to use `#[global_allocator]` is to choose it when you are compiling a Rust binary. Putting it in libraries is a bad idea, except probably for libraries like `jemallocator-global`, whose only job is to set a particular allocator as the `#[global_allocator]`, so that a binary only has to include that crate as a dependency to pick the allocator up, without having to write any code.

**Comment by glandium** [comment]
But isn't the allocator you get when there isn't a #[global_allocator] set via an implicit #[global_allocator]? Here's what happens when I take the test case from this crate and remove the #[global_allocator] from the foo crate: foobar outputs nothing (default allocator) and barfoo prints bar twice. So if two #[global_allocator] should conflict, whatever they are, then linking two staticlibs when none of them have one would too?

**Comment by gnzlbg** [comment]
> But isn't the allocator you get when there isn't a #[global_allocator] set via an implicit #[global_allocator]? 

If a **binary** doesn't contain a user-defined `#[global_alloctor]`, then the compiler inserts one for you that calls the system allocator. A library is not a binary. You can have two static libraries, and link them together into an executable binary. It is up to the final binary to choose the `#[global_allocator]`.

**Comment by glandium** [comment]
> It is up to the final binary to choose the #[global_allocator].

The final linker doesn't know anything about global allocator in this test case.

**Comment by glandium** [comment]
As in, static libraries are different. They *do* contain the global allocator that calls the system allocator.

**Comment by gnzlbg** [comment]
> The final linker doesn't know anything about global allocator in this test case.

The linker doesn't have to. When rustc compiles an executable binary linking static libraries to it, it can check whether any of those libraries contain a global-allocator. If they don't, and if the binary doesn't either, then and only then it is sound to generate a #[global_allocator].

> As in, static libraries are different. They do contain the global allocator that calls the system allocator.

That sounds like a bug. This would mean that you can't link a static library to anything that defines `#[global_allocator]` without undefined behavior (e.g. without one of the two allocators being picked, depending on link order).

**Comment by glandium** [comment]
No, it's perfectly normal. How else would you link rust static libraries to non-rust things? In your assumed model, rustc is always compiling/linking the final binary. That's not always the case. See the very test-case in this crate.

**Comment by glandium** [comment]
> This would mean that you can't link a static library to anything that defines #[global_allocator] without undefined behavior (e.g. without one of the two allocators being picked, depending on link order).

Note, that's exactly what the test-case in this crate shows.

**Comment by glandium** [comment]
But your suggested fix would break the case where neither of the static libraries defines a global-allocator, because the rust compiler does implicitly define one.

**Comment by gnzlbg** [comment]
> No, it's perfectly normal. How else would you link rust static libraries to non-rust things?

Those non-rust things need to link a global allocator, e.g., the system one.

**Comment by gnzlbg** [comment]
> But your suggested fix would break the case where neither of the static libraries defines a global-allocator, because the rust compiler does implicitly define one.

That would compile fine, you just need to also link a global allocator when using them.

**Comment by glandium** [comment]
> Those non-rust things need to link a global allocator, e.g., the system one.

They need to link malloc/free, not a global allocator. The rust global allocator is in the rust static library.

**Comment by gnzlbg** [comment]
> They need to link malloc/free, not a global allocator. The rust global allocator is in the rust static library.

The global allocator is just 4 symbols, if those aren't present, linking should fail. 

> The rust global allocator is in the rust static library.

The global allocator doesn't need to be in the static library for that to work.

**Comment by glandium** [comment]
> The global allocator doesn't need to be in the static library for that to work.

The point is it *is* in the static library! As well as libstd, libcore, etc. How else is the final link supposed to get all these? Manually? That would also be a breaking change if they stopped being there.

**Comment by gnzlbg** [comment]
> The point is it is in the static library!

No it isn't. Even if you were to link Rust `#[global_allocator]` implementation for it, this code actually calls symbols (e.g. `malloc`) that aren't part of the static library. 

> As well as libstd, libcore, etc. How else is the final link supposed to get all these?

If you check these out, they all call symbols that aren't in the library, e.g. `f32.sin()` calls `_sinf`, which isn't anywhere there. 

> Manually?

No, **you can't execute a library**. It doesn't matter _at all_ whether the library has all symbols it uses as long as the final binary has them. The final binary links them, e.g., via a dynamic linker, or by providing them as part of the binar, but it is up to the final binary to choose which math library, allocator, etc. are used.

**Comment by gnzlbg** [comment]
> No, you can't execute a library. It doesn't matter at all whether the library has all symbols it uses as long as the final binary has them. The final binary links them, e.g., via a dynamic linker, or by providing them as part of the binar, but it is up to the final binary to choose which math library, allocator, etc. are used.

You can try to statically link as much code as possible, but then using your library with others that also statically link such code invokes undefined behavior.

**Comment by glandium** [comment]
We're definitely not talking about the same thing. malloc/free is what the global-allocator implementation calls. It's irrelevant with what I'm talking about, which is the global-allocator implementation itself. And now that I've looked, it's even more subtle.
Each rust static library has `__rust_alloc`, `__rust_dealloc`, `__rust_alloc_zeroed` and `__rust_realloc` symbols. That's what all code doing allocations end up calling.
What /those/ functions call depends on whether the crate has a #[global-allocator] or not. In the case it does, `__rg_alloc`, `__rg_dealloc`, `__rg_alloc_zeroes`, `__rg_realloc` are called, respectively. In the case it doesn't, `__rd_alloc`, `__rd_dealloc`, `__rd_alloc_zeroes`, `__rd_realloc` are called.
The `__rg_*` functions are essentially the GlobalAlloc implementation if there is one, and the `_rd_*` functions are functions that call `malloc`/`free`.
Now, what happens in this test case is that when the linker wants to find the `__rust_*` functions, it finds one or the other set, depending on which came first, which then resolve to calls to one set of `__rg_*` or `__rd_*`, and the other functions are dropped.
I don't know how you'd want to alter this such that linking 2 static libraries without a #[global-allocator] (so, all linking `__rust_*` and `__rd_*`) still works while making hybrids setup fail to link (one crate with `__rg_*`, one with `__rd_*` or both crates with different `__rg_*`)

**Comment by gnzlbg** [comment]
> And now that I've looked, it's even more subtle.

Finally.

> Each rust static library has __rust_alloc, __rust_dealloc, __rust_alloc_zeroed and __rust_realloc symbols. That's what all code doing allocations end up calling.

All code doing allocations calls those symbols. 

Those symbols should be provided by rustc when generating a binary, a static library iff the user provided a `#[global_allocator]` in the static library, etc.

In the cases in which rustc should provide these symbols, it does so by forwarding them to call the `__rg_` or `__rd_` prefixed ones, depending on whether a `#[global_allocator]` was provided by the user or not. 

So if you have a static library, and it does not provide a `#[global_allocator]`, that library calls `__rust_alloc`, but that symbol does not exist. If you have another static library, and this library provides a `#[global_allocator]`, that library calls `__rust_alloc`, and the `__rust_alloc` symbol is defined there to call the `__rg_` symbols of the global allocator. If you then go and link both libraries, everything works. If you then link the result of that into a Rust binary, then everything works too because Rust should only define those for a binary if there is no user-defined global allocator, but there is one in one of the static libraries. 

If you create a **Rust** static library (staticlib, not cdylib) that allocates memory and does not provide a `#[global_allocator]`, and then link it with C code, then linking with fail because the `__rust_alloc` symbol does not exist. If you instead compile that library as a `cdylib`, then Rust will always define `__rust_alloc` and all the definitions will be identical and call the `__rd_` symbols of the system allocator in liballoc, which will call `malloc` and friends. The `malloc` and friends symbols are not provided anywhere. So if you link your cdylib with a C binary that doesn't link `malloc`, then it should fail to link. If you link your cdylib with a C binary or library that links the C standard library, that library provides malloc and everything works.

>  don't know how you'd want to alter this such that linking 2 static libraries without a #[global-allocator] (so, all linking __rust_* and __rd_*)

Those two Rust `staticlib` libraries should **use** the `__rust_` symbols, but they should not declare them. That is, they should not contain any `__rd_` or `__rg_` symbols. 

Only if you make both libraries a `cdylib` should they both define `__rust_` and `__rd_` and have `__rust_` call `__rd_` which will use `malloc`. These symbols should be marked as "mergeable" for cdylibs, and picking one or the other does not result in undefined behavior because both are exactly identical.

**Comment by gnzlbg** [comment]
> That is, they should not contain any __rd_ or __rg_ symbols.

Oops, this was wrong. Whether the `__rd_` symbols are part of the library depends on whether the library links `liballoc`, which contains a global allocator with those symbols. That's ok, because those symbols are identical for all libraries linking liballoc. 

Note that this `__rd_` symbols do not need to be used by anything though. For that case of the two Rust staticlib, no definition of the `__rust_` symbols should exist, so nothing calls the `__rd_` symbols. That's ok.

**Comment by glandium** [comment]
> > And now that I've looked, it's even more subtle.
>
> Finally.

I'll return you the favor. You obviously haven't looked at a rust static library, with or without #[global-allocator].

> Those two Rust staticlib libraries should use the __rust_ symbols, but they should not declare them. That is, they should not contain any __rd_ or __rg_ symbols.

THEY DO! And we're now back to https://github.com/glandium/global_allocator_linkage-rs/issues/1#issuecomment-520775858

**Comment by gnzlbg** [comment]
> You obviously haven't looked at a rust static library, with or without #[global-allocator].

Why do you say that?

> THEY DO! A

I know that they do, the whole point of this thread is whether that is a bug. I think it is.
